### PR TITLE
Added type hints to the Column class

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes

### CODE REVIEW

File: lib/Doctrine/DBAL/Schema/Column.php [hunk 1]
```diff
@@ -21,13 +20,13 @@ class Column extends AbstractAsset
     protected $_type;
 
     /** @var int|null */
-    protected $_length = null;
+    protected $_length;
 
     /** @var int */
-    protected $_precision = 10;
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Schema/Column.php]: Not sure if this should be changed

File: lib/Doctrine/DBAL/Schema/Column.php [hunk 2]
```diff
@@ -21,13 +20,13 @@ class Column extends AbstractAsset
...
+    protected $_precision;
 
     /** @var int */
-    protected $_scale = 0;
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Schema/Column.php]: Same here: default value to be preserved IMO

**Review by Ocramius — DISMISSED** [review]

File: lib/Doctrine/DBAL/Schema/Column.php (hunk 1 above)
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/Column.php]: As precision is only applicable for numerical columns, I think it does make sense to drop the default.

Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/Column.php]: Another solution could be moving this information to Type.

File: lib/Doctrine/DBAL/Schema/Column.php [hunk 3]
```diff
@@ -94,106 +90,59 @@ public function setOptions(array $options)
...
-     */
-    public function setPrecision($precision)
+    public function setPrecision(?int $precision) : self
     {
-        if (! is_numeric($precision)) {
-            $precision = 10; // defaults to 10 when no valid precision is given.
-        }
-
-        $this->_precision = (int) $precision;
+        $this->_precision = $precision;
```
Comment (Majkl578) [inline lib/Doctrine/DBAL/Schema/Column.php]: Should we check for `$precision >= 0`?

File: lib/Doctrine/DBAL/Schema/Column.php [hunk 4]
```diff
@@ -94,106 +90,59 @@ public function setOptions(array $options)
...
-     */
-    public function setScale($scale)
+    public function setScale(?int $scale) : self
     {
-        if (! is_numeric($scale)) {
-            $scale = 0;
-        }
-
-        $this->_scale = (int) $scale;
+        $this->_scale = $scale;
```
Comment (Majkl578) [inline lib/Doctrine/DBAL/Schema/Column.php]: Should we check for `$scale >= 0`?

File: lib/Doctrine/DBAL/Schema/Column.php [hunk 5]
```diff
@@ -94,106 +90,59 @@ public function setOptions(array $options)
...
-     */
-    public function setLength($length)
+    public function setLength(?int $length) : self
     {
-        if ($length !== null) {
-            $this->_length = (int) $length;
-        } else {
-            $this->_length = null;
-        }
+        $this->_length = $length;
```
Comment (Majkl578) [inline lib/Doctrine/DBAL/Schema/Column.php]: Should we check for `$length > 0`?

File: lib/Doctrine/DBAL/Schema/Column.php (hunk 1 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Column.php]: This is a DTO which just carries what's given to it. It shouldn't imply any defaults. Whatever owns the definition (DB schema or the definition in the application) should define it explicitly.

I'll add it to the upgrade notes.

File: lib/Doctrine/DBAL/Schema/Column.php (hunk 3 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Column.php]: I don't think so:

1. It's out of scope.
2. The class is just a DTO, not a domain object. DBAL doesn't do any handling of the precision AFAIK, and the DB will validate it itself.
3. In the domain, the value is not only bounded from at bottom (is it?), it also may be bounded at the top which we know nothing about at the abstraction layer.
4. The max value may be dependent on other values (e.g. on Oracle, the sum of precision and scale can't be greater than 38).

**Comment by jwage** [comment]
Can we (should we?) separate the BC breaks due to adding type hints and BC breaks from us changing default behavior? IMO, while changing the defaults may be a good idea long term, I am not sure if making the change along side adding type hints is good. May be better to have 1 PR to add the type hints while maintaining behavior and then another PR to change the defaults. This way the defaults changing can be discussed in isolation from all the other changes and the type hints can probably be merged faster.

**Comment by morozov** [comment]
@jwage I get the intent but if we want to take this route, the changes have to be implemented in the reverse order:

1. Drop the defaults (the real BC break)
2. Enforce strict argument types (a break only for API implementors, not consumers).

The existing defaults only make sense with non-enforced argument types (hence the `is_numeric()` checks). If we introduce the strict types first, it will implicitly break the behavior covered by `is_numeric()`.

I like the idea of the split in general. Especially, given that there are still some leftovers of the default value of `10` in `Comparator`. I'll open a separate PR.

File: lib/Doctrine/DBAL/Schema/Column.php (hunk 2 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Column.php]: Reverted from this patch.

File: lib/Doctrine/DBAL/Schema/Column.php (hunk 1 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Column.php]: Reverted from this patch.

**Review by jwage — APPROVED** [review]

**Pull request merged**